cpp/src/bitmask/bit_mask.h

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+   * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef BIT_MASK_H
+#define BIT_MASK_H
+
+#include "utilities/cudf_utils.h"
+
+namespace bit_mask {
+  typedef uint32_t     bit_mask_t;
+  
+  namespace detail {
+    constexpr gdf_size_type BITS_PER_ELEMENT{sizeof(bit_mask_t) * 8};
+
+    /**
+     *  @brief determine the bit mask element that contains a record
+     *
+     *  @param[in]  record_idx    The record index
+     *
+     *  @return the bit mask element index
+     */
+    template <typename T>
+    CUDA_HOST_DEVICE_CALLABLE
+    constexpr gdf_size_type which_element(T record_idx) {
+      return (record_idx / BITS_PER_ELEMENT);
+    }
+
+    /**
+     *  @brief determine which bit in a bit mask element relates to a record
+     *
+     *  @param[in]  record_idx    The record index
+     *
+     *  @return which bit within the bit mask element
+     */
+    template <typename T>
+    CUDA_HOST_DEVICE_CALLABLE
+    constexpr gdf_size_type which_bit(T record_idx) {
+      return (record_idx % BITS_PER_ELEMENT);
+    }
+  }
+
+  /**
+   * @brief determine the number of bit_mask_t elements are used
+   *
+   * @param[in]  size    Number of bits in the bitmask
+   *
+   * @return the number of elements
+   */
+  CUDA_HOST_DEVICE_CALLABLE
+  constexpr gdf_size_type num_elements(gdf_size_type size) { 
+    return (( size + ( detail::BITS_PER_ELEMENT - 1)) / detail::BITS_PER_ELEMENT ); 
+  }
+
+  /**
+   *  @brief Copy data between host and device
+   *
+   *  @param[out] dst      - the address of the destination
+   *  @param[in] src       - the address of the source
+   *  @param[in] num_bits  - the number of bits in the bit container
+   *  @param[in] kind      - the direction of the copy
+   *
+   *  @return GDF_SUCCESS on success, the CUDA error on failure
+   */
+  inline gdf_error copy_bit_mask(bit_mask_t *dst, const bit_mask_t *src, size_t num_bits, enum cudaMemcpyKind kind) {
```

Comment:
we should really start using streams here. If we don't these memcpy can slow down everyone else. can we start using cudaMemcpyAsync with streams to avoid serializing things?

Reply:
How often do we expect to copy an entire bit map from host to device or back?  I'm not sure of all of the bit mask use cases.  I simply created a function to replace existing code which did the memcpy.

Reply:
Let's get this PR in and file an issue to make the copy async. Can always default to null stream. BTW, you should also be able to eliminate kind by passing cudaMemcpyDefault as kind to the underlying memcpy call...  UVA should work on all GPUs that RAPIDS supports...